app/presenters/content_embed_presenter.rb

```diff
@@ -0,0 +1,54 @@
+module Presenters
+  class ContentEmbedPresenter
+    def initialize(edition)
+      @edition = edition
+    end
+
+    def render_embedded_content(details)
+      return details if details[:body].nil?
+
+      details[:body] = if details[:body].is_a?(Array)
+                         details[:body].map do |content|
+                           {
+                             content_type: content[:content_type],
+                             content: render_embedded_editions(content[:content]),
+                           }
+                         end
+                       else
+                         render_embedded_editions(details[:body])
+                       end
+
+      details
+    end
+
+  private
+
+    def render_embedded_editions(content)
+      embedded_content_references = EmbeddedContentFinderService.new.find_content_references(content)
+      embedded_content_references_by_content_id = embedded_content_references.index_by(&:content_id)
+
+      target_content_ids = @edition
+        .links
+        .where(link_type: "embed")
+        .map(&:target_content_id)
+
+      embedded_edition_ids = ::Queries::GetEditionIdsWithFallbacks.call(
+        target_content_ids,
+        locale_fallback_order: [@edition.locale, Edition::DEFAULT_LOCALE],
+        state_fallback_order: %w[published],
+      )
+
+      embedded_editions = Edition.where(id: embedded_edition_ids)
+
+      embedded_editions.each do |embedded_edition|
+        embed_code = embedded_content_references_by_content_id[embedded_edition.content_id].embed_code
+        content = content.gsub(
+          embed_code,
+          embedded_edition.title,
```

Comment:
Ah, I think this is actually not the right field on contact to be replacing things with. `title` will be something like "Government Digital Service", rather than "10 Whitechapel High Street" like we probably want.

The way Whitehall does it is:

- [Render the contacts/contact partial](https://github.com/alphagov/whitehall/blob/81beec81ef99e1c8d230f028a8b725e406770371/app/helpers/govspeak_helper.rb#L160-L161)
- Which [is actually proper complicated](https://github.com/alphagov/whitehall/blob/main/app/views/contacts/_contact.html.erb)
- Complicated further by the fact that [the code to render addresses](https://github.com/alphagov/whitehall/blob/main/app/helpers/admin/contacts_helper.rb#L2) [is itself very complicated](https://github.com/alphagov/whitehall/blob/main/lib/address_formatter/formatter.rb#L3)

🤔  we could merge this with it just using title, but it's probably not actually what we want. On the other hand, I don't think publishing-api having a bunch of knowledge of how to render addresses is desirable - I'd be somewhat resistant to copying the templating code over from Whitehall.

I think the simplest option I can think of would be to add an "embeddable_html" field to Contact when Whitehall presents Contacts to publishing-api, and then use that. That way Whitehall can continue doing the rendering of Contact HTML (even though that's not very nice either), and publishing-api doesn't need to get rendering app capabilities.

Longer term with the design of this feature, I wonder if this is a case for passing the `{{embed:...}}` syntax all the way through to the frontend, and having the rendering app do the rendering of contact data into HTML (having followed the expanded link).

I'm sort of leaning towards getting this PR merged and then fixing this in a subsequent PR though.

Reply:
We'll leave this as `title` for now (on the basis that no-one will actually be embedding anything into live content) but will write up a card to implement `embedded_html` in the schemas and include that in the document instead.

>Longer term with the design of this feature, I wonder if this is a case for passing the {{embed:...}} syntax all the way through to the frontend, and having the rendering app do the rendering of contact data into HTML (having followed the expanded link).

This makes a lot of sense to me.  Particularly as the frontend app is better placed to do any conversion to HTML (rather than the current situation of the publishing app or Publishing API doing this, then any frontend changes needing all content to be republished).